query/search.js

```diff
@@ -33,50 +32,68 @@ var SQL = [
   'WHERE rowid IN (',
     'SELECT rowid FROM (',
       'SELECT * FROM base',
-      'WHERE housenumber < "%%TARGET_HOUSENUMBER%%"',
+      'WHERE housenumber < $housenumber',
       'GROUP BY id HAVING( MAX( housenumber ) )',
       'ORDER BY housenumber DESC',
     ')',
     'UNION',
     'SELECT rowid FROM (',
       'SELECT * FROM base',
-      'WHERE housenumber >= "%%TARGET_HOUSENUMBER%%"',
+      'WHERE housenumber >= $housenumber',
       'GROUP BY id HAVING( MIN( housenumber ) )',
       'ORDER BY housenumber ASC',
     ')',
   ')',
   'ORDER BY housenumber ASC', // @warning business logic depends on this
-  'LIMIT %%MAX_MATCHES%%;'
+  `LIMIT ${MAX_MATCHES};`
 ].join(' ');
 
-var NAME_SQL = '(street.names.name=?)';
+// SQL prepared statements dont easily support variable length inputs.
+// This function dynamically generates a SQL query based on the number
+// of 'name' conditions required.
+function generateDynamicSQL(max){
+  const conditions = new Array(max.names)
+    .fill('(street.names.name=$name)')
+    .map((sql, pos) => sql.replace('$name', `$name${pos}`));
+
+  return SQL.replace('%%NAME_CONDITIONS%%', conditions.join(' OR '));
+}
 
-module.exports = function( db, point, number, names, cb ){
+// Reusing prepared statements can have a ~10% perf benefit
+// Note: the cache is global and so must be unique per database.
+const cache = [];
+function statementCache(db, max){
```

Comment:
why is "max" a dict with a nested integer, instead of just an int? would be much easier to understand